text/0123-improve-span-status-api.md

```diff
@@ -0,0 +1,114 @@
+
+- Logging isn't a simple or cheap dependency to take in the implementation. While OpenTelemetry may solidify a logging offering for API and SDK components, an end-to-end scenario still requires snooping the log stream to determine relevant messages for a given span. Regardless whether this is done client-side, at the database or in the UI layer it potentially involves performance overhead of handling an order of magnitude more data.
+- It requires establishing conventions that designate which log message represents 'the status' for a Span rather than one of potentially many results or errors that were recorded during the Span's duration. This likely means this status is a special case on the logging API if it isn't a special case on the distributed trace API.
+- I expect developers both at the time they are emitting trace data and when viewing that trace data would find it is idiomatic to  include the status of the Span's workload together with the description of the Span. Identifying failed or abnormal spans is a typical APM operation.
+
+**Closed vs. open-ended status descriptions** - We could make status represented by a fixed set of options such as the current gRPC codes or any larger/different set. It likely is simpler for tools consuming the data to only operate on fixed options and it is pleasant to imagine that all (or most) status results will be reasonably categorized into the fixed list. However I do not expect most users to take up the responsibility of mapping their status results if OpenTelemetry/back-end tools don't provide a clear representation. Instead these users will not set status or choose options that are generic and inconsistent rendering the data useless to everyone. For a fixed list to work I believe we either have to curate a very large list of options or accept that significant portions of the developer audience will not find the status representation useful.
+
+**API using attribute semantic conventions ** - It is also possible to do this via semantic conventions on attributes and although I think the strongly typed API is preferable I don't have that strong of an opinion. Semantic conventions are likely to have higher performance overhead, higher risk of error in key names and are less discoverable/refactorable in IDEs. The advantages are that there is some past precedent for doing this specific to http codes and new conventions can be added easily. If we go semantic conventions it does imply that Exception becomes a type that can be directly passed as an argument to SetAttribute(). Requiring the user to destructure the exception into a list of key value pairs would be overly onerous and error-prone for a common usage scenario. If desired the SDK or exporter could destructure it, but that can be determined independently from API design and I'd like to keep it out of scope.
```

Comment:
Note that you can provide a nicely typed API and still use semantic conventions. E.g. the API could just "serialize" whatever you give it into SetAttribute / CreateEvent calls. The difference is that the API between the exporter and the API is then "weakly typed".

To point at another project, the idea of "typed spans", i.e. utilities to have strongly typed APIs for semantic conventions, is at least as old as OpenTelemetry. E.g. see "Typed Spans" in the PR description of https://github.com/open-telemetry/opentelemetry-specification/pull/571

Reply:
As an additional advantage, the exception APIs could be provided as a package separate from the API, (you could provide a function `com.opentelemetry.contrib.errorutils.ErrorUtils.captureError(Span, other params)` that sets any attributes/events on the span).

Reply:
Thanks! I've added this suggestion to the list of alternatives.